Refactor File types to have common functionality use traits

josephlr opened the pull request:
In `proto::media::file`, `File` and `Directory` share most of their functionality in common. In general, such common functionality is easier to deal with in rust though traits.

This change refactors the existing code into 3 structs and 1 trait
- `File` a trait containing all the functionality common to all file handles.
- `FileHandle` a basic file handle struct, allowing for conversion to the more specific structs. This is what `open()` returns.
- `RegularFile` a normal data file, allowing for direct read/write.
- `Directory` a directory file, allowing for enumerating its contents. This is returned by `open_volume()`.

There are three main questions I have about this change.
1. There isn't really any reason to have a lifetime bound on `File`, `Directory`, etc... Should I change the contained reference to `FileImpl` to be a pointer or `'static` reference? It would make the external API nicer.
2. There is a lot of repeated `self.file().0` throughout the code now. Should I add an internal helper method that just take a `File` and outputs a `&mut FileImpl`?
3. Naming of the structs/trait. I went with `File - FileHandle - Directory/RegualrFile`, but we could also use names like:
  a. `Handle`
  b. `Object`
  c. `FileTrait`
  d. `DataFile`
  e. `Base`

GabrielMajeri commented on src/proto/media/file/regular.rs:
```diff
@@ -0,0 +1,99 @@
+    ///
+    /// Seeking past the end of the file is allowed, it will trigger file growth on the next write.
+    ///
+    /// The special value 0xFFFF_FFFF_FFFF_FFFF may be used to seek to the end of the file.
```
Could we get add this as an associated constant in the impl block for `RegularFile`?

GabrielMajeri commented on src/proto/media/file/mod.rs:
```diff
@@ -16,32 +18,24 @@ use core::mem;
-    pub(super) unsafe fn new(ptr: *mut FileImpl) -> Self {
-        File(&mut *ptr)
-    }
+/// `FilesystemObject` continas all funcationality that is safe to perform on
```
Small typos here

GabrielMajeri commented on src/proto/media/file/regular.rs:
```diff
@@ -0,0 +1,99 @@
+    /// Get the file's current position
+    ///
+    /// # Errors
+    /// * `uefi::Status::UNSUPPORTED`    Attempted to get the position of an opened directory
```
Probably can remove this from the list if we statically know this is a regular file.

josephlr commented:
> The code repetition for `self.file().0` is a bit annoying, any improvement you can do is appreciated.

I added a `FileInternal` helper trait with a single `imp()` method that gives access to the function pointer table. This will be useful if we ever want to add other internal helper methods.
 
> I don't have any better ideas for the naming of the structs, but I think the way you named them makes their usage unambiguous.

After giving the docs a test read, I made a few changes to make the API more readable. Let me know what you think:
- `File` is the trait
- `FileHandle` is the basic struct
- `Directory` and `RegularFile` are the specific structs
- `FileType` is the enum that returned from `into_type()`
- Unsafe construction of `Directory` and `RegularFile` is now done though a `new` method.

I also rebased on #87 so that Travis CI will be happy.

josephlr commented:
@GabrielMajeri I was also now able to make `FileHandle` just contain a pointer, so now there are no generic lifetime params for any of the structs/traits.

josephlr commented:
@GabrielMajeri looks like we tried the exact same fix (rebase/force-push) at the exact same time, it looks like everything worked though.

GabrielMajeri commented on src/proto/media/file/info.rs:
```diff
@@ -9,7 +9,7 @@ use core::mem;
 
 /// Common trait for data structures that can be used with
-/// `File::set_info()` or `File::set_info()`.
+/// `FilesystemObject::set_info()` or `FilesystemObject::get_info()`.
```
I think this now refers to `File`

josephlr replied:
Done
